### PR TITLE
add block buffer for block chain and update unit test

### PR DESCRIPTION
**Pull request by lichenxi-webank**
imporve performance by avoiding reading database repeatedly when call getBlockByHash

### CODE REVIEW

File: libblockchain/BlockChainImp.cpp [hunk 1]
```diff
@@ -43,6 +43,56 @@ using namespace dev::blockverifier;
...
+std::shared_ptr<Block> BlockCache::add(Block& _block)
+{
+    BLOCKCHAIN_LOG(TRACE) << "[#add] Add block to block cache, [blockHash]: "
+                          << _block.header().hash();
+
+    {
+        WriteGuard guard(m_sharedMutex);
+        if (m_blockCache.size() > c_blockCacheSize)
+        {
+            BLOCKCACHE_LOG(INFO) << "[#add] Block cache full, start to remove old item...";
```
Comment (wheatli) [inline libblockchain/BlockChainImp.cpp]: INFO -> DEBUG/TRACE

File: libblockchain/BlockChainImp.h [hunk 2]
```diff
@@ -48,6 +53,23 @@ class MemoryTableFactory;
...
+    BlockCache(){};
+    std::shared_ptr<dev::eth::Block> add(dev::eth::Block& _block);
+    std::pair<std::shared_ptr<dev::eth::Block>, dev::h256> get(h256 const& _hash);
+
+private:
+    mutable boost::shared_mutex m_sharedMutex;
+    mutable std::map<dev::h256, std::pair<std::shared_ptr<dev::eth::Block>, dev::h256> >
+        m_blockCache;
+    mutable std::deque<dev::h256> m_blockCacheFIFO;  // insert queue log for m_blockCache
+    const unsigned c_blockCacheSize = 1024;          // m_blockCache size, default set 1024
```
Comment (wheatli) [inline libblockchain/BlockChainImp.h]: It's too large

File: libblockchain/BlockChainImp.cpp (hunk 1 above)
Comment (lichenxi-webank) [reply libblockchain/BlockChainImp.cpp]: Done.

File: libblockchain/BlockChainImp.h (hunk 2 above)
Comment (lichenxi-webank) [reply libblockchain/BlockChainImp.h]: Now the cache size has changed to 10.

**Pull request merged**